crates/wasm-smith/src/lib.rs

```diff
@@ -1476,60 +1476,32 @@ where
             return Ok(());
         }
 
-        let mut choices: Vec<fn(&mut Unstructured, &mut ConfiguredModule<C>) -> Result<_>> =
-            Vec::with_capacity(4);
-
-        if self.funcs.len() > 0 {
-            choices.push(|u, m| {
-                let idx = u.int_in_range(0..=m.funcs.len() - 1)?;
-                let sig = &m.funcs[idx].1;
-                m.type_size += 1 + (sig.params.len() + sig.results.len()) as u32;
-                Ok(Export::Func(idx as u32))
-            });
-        }
-
-        if self.tables.len() > 0 {
-            choices.push(|u, m| {
-                let idx = u.int_in_range(0..=m.tables.len() - 1)?;
-                m.type_size += 1;
-                Ok(Export::Table(idx as u32))
-            });
+        let mut choices = Vec::with_capacity(
+            self.funcs.len()
+                + self.tables.len()
+                + self.memories.len()
+                + self.globals.len()
+                + self.instances.len()
+                + self.modules.len(),
+        );
+
+        for i in 0..self.funcs.len() {
+            choices.push(Export::Func(i as u32));
```

Comment:
This is making the size of choices grow from `O(1)` to `O(n)` since we are pushing each function onto `choices` rather than doing a single push for all functions, and similar for other entities.

Can we do the approach we had before, except we only push the export-a-function choice if there exist any functions that won't blow the type size limit, and then do the filtering of which functions are available only if the choice to export a function was made?

Reply:
Sure! I've pushed up a version which does that now